Compile error with G4 10.6.3

Celeritas no longer compiles with Geant4 10.6.3 as its using API function that were added in a later version. The following files failed to compile: 

```
FAILED: src/celeritas/CMakeFiles/celeritas_g4vg.dir/ext/g4vg/SolidConverter.cc.o 
/cvmfs/sft.cern.ch/lcg/releases/gcc/11.3.0-ad0f5/x86_64-centos7/bin/g++ -DG4LIB_BUILD_DLL -DVECCORE_ENABLE_CUDA -I/home/esseivaj/bld4/devel/celeritas/src -I/home/esseivaj/bld4/devel/celeritas/build/include -isystem /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/include -isystem /opt/cuda/11.8.0/include -isystem /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/include/Geant4 -isystem /home/esseivaj/bld4/devel/spack/opt/spack/linux-centos7-cascadelake/gcc-11.3.0/clhep-2.4.6.0-cxuvt2z7pmefdpbcr4jzgu2qkhd3xsxe/include -isystem /bld4/home/esseivaj/devel/spack/opt/spack/linux-centos7-cascadelake/gcc-11.3.0/zlib-1.2.13-yaqekz3x2qihs3digztjtq3eahggv6t4/include -isystem /bld4/home/esseivaj/devel/spack/opt/spack/linux-centos7-cascadelake/gcc-11.3.0/expat-2.5.0-gzduk2bfkxfqoxwkqs6fxkxqqi2fe2bq/include -isystem /bld4/home/esseivaj/devel/spack/opt/spack/linux-centos7-cascadelake/gcc-11.3.0/xerces-c-3.2.3-4yis4cphgpkcynfxir4346muuwnvgds3/include -Wall -Wextra -Wno-psabi -pedantic -pedantic-errors -g -fPIC -faligned-new -std=c++17 -MD -MT src/celeritas/CMakeFiles/celeritas_g4vg.dir/ext/g4vg/SolidConverter.cc.o -MF src/celeritas/CMakeFiles/celeritas_g4vg.dir/ext/g4vg/SolidConverter.cc.o.d -o src/celeritas/CMakeFiles/celeritas_g4vg.dir/ext/g4vg/SolidConverter.cc.o -c /home/esseivaj/bld4/devel/celeritas/src/celeritas/ext/g4vg/SolidConverter.cc
/home/esseivaj/bld4/devel/celeritas/src/celeritas/ext/g4vg/SolidConverter.cc: In member function 'vecgeom::cxx::VUnplacedVolume* celeritas::g4vg::SolidConverter::para(celeritas::g4vg::SolidConverter::arg_type)':
/home/esseivaj/bld4/devel/celeritas/src/celeritas/ext/g4vg/SolidConverter.cc:393:15: error: 'const class G4Para' has no member named 'GetTheta'
  393 |         solid.GetTheta(),
      |               ^~~~~~~~
/home/esseivaj/bld4/devel/celeritas/src/celeritas/ext/g4vg/SolidConverter.cc:394:15: error: 'const class G4Para' has no member named 'GetPhi'
  394 |         solid.GetPhi());
      |               ^~~~~~
/home/esseivaj/bld4/devel/celeritas/src/celeritas/ext/g4vg/SolidConverter.cc: In member function 'vecgeom::cxx::VUnplacedVolume* celeritas::g4vg::SolidConverter::trap(celeritas::g4vg::SolidConverter::arg_type)':
/home/esseivaj/bld4/devel/celeritas/src/celeritas/ext/g4vg/SolidConverter.cc:566:15: error: 'const class G4Trap' has no member named 'GetTheta'
  566 |         solid.GetTheta(),
      |               ^~~~~~~~
/home/esseivaj/bld4/devel/celeritas/src/celeritas/ext/g4vg/SolidConverter.cc:567:15: error: 'const class G4Trap' has no member named 'GetPhi'
  567 |         solid.GetPhi(),
      |               ^~~~~~
/home/esseivaj/bld4/devel/celeritas/src/celeritas/ext/g4vg/SolidConverter.cc:571:15: error: 'const class G4Trap' has no member named 'GetAlpha1'; did you mean 'GetTanAlpha1'?
  571 |         solid.GetAlpha1(),
      |               ^~~~~~~~~
      |               GetTanAlpha1
/home/esseivaj/bld4/devel/celeritas/src/celeritas/ext/g4vg/SolidConverter.cc:575:15: error: 'const class G4Trap' has no member named 'GetAlpha2'; did you mean 'GetTanAlpha2'?
  575 |         solid.GetAlpha2());
      |               ^~~~~~~~~
      |               GetTanAlpha2
FAILED: src/accel/CMakeFiles/accel_objects.dir/SetupOptionsMessenger.cc.o 
/cvmfs/sft.cern.ch/lcg/releases/gcc/11.3.0-ad0f5/x86_64-centos7/bin/g++ -DG4LIB_BUILD_DLL -DVECCORE_ENABLE_CUDA -I/home/esseivaj/bld4/devel/celeritas/src -I/home/esseivaj/bld4/devel/celeritas/build/include -isystem /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/include -isystem /opt/cuda/11.8.0/include -isystem /home/esseivaj/bld4/devel/spack/var/spack/environments/celeritas/.spack-env/view/include/Geant4 -isystem /home/esseivaj/bld4/devel/spack/opt/spack/linux-centos7-cascadelake/gcc-11.3.0/clhep-2.4.6.0-cxuvt2z7pmefdpbcr4jzgu2qkhd3xsxe/include -isystem /bld4/home/esseivaj/devel/spack/opt/spack/linux-centos7-cascadelake/gcc-11.3.0/zlib-1.2.13-yaqekz3x2qihs3digztjtq3eahggv6t4/include -isystem /bld4/home/esseivaj/devel/spack/opt/spack/linux-centos7-cascadelake/gcc-11.3.0/expat-2.5.0-gzduk2bfkxfqoxwkqs6fxkxqqi2fe2bq/include -isystem /bld4/home/esseivaj/devel/spack/opt/spack/linux-centos7-cascadelake/gcc-11.3.0/xerces-c-3.2.3-4yis4cphgpkcynfxir4346muuwnvgds3/include -Wall -Wextra -Wno-psabi -pedantic -pedantic-errors -g -fPIC -faligned-new -std=c++17 -MD -MT src/accel/CMakeFiles/accel_objects.dir/SetupOptionsMessenger.cc.o -MF src/accel/CMakeFiles/accel_objects.dir/SetupOptionsMessenger.cc.o.d -o src/accel/CMakeFiles/accel_objects.dir/SetupOptionsMessenger.cc.o -c /home/esseivaj/bld4/devel/celeritas/src/accel/SetupOptionsMessenger.cc
/home/esseivaj/bld4/devel/celeritas/src/accel/SetupOptionsMessenger.cc: In static member function 'static long int celeritas::{anonymous}::UICommandTraits<T, typename std::enable_if<is_integral_v<T>, void>::type>::from_string(const G4String&)':
/home/esseivaj/bld4/devel/celeritas/src/accel/SetupOptionsMessenger.cc:57:29: error: 'ConvertToLongInt' is not a member of 'G4UIcommand'
   57 |         return G4UIcommand::ConvertToLongInt(v.c_str());
      |                             ^~~~~~~~~~~~~~~~
```

--- a/src/accel/SetupOptionsMessenger.cc
+++ b/src/accel/SetupOptionsMessenger.cc
@@ -56,13 +56,13 @@ struct UICommandTraits<T, std::enable_if_t<std::is_integral_v<T>>>
     static long from_string(G4String const& v)
     {
         // Conversion to long int introduced in Geant4 10.7.0
-#if G4VERSION_NUMBER < 1070
+#if G4VERSION_NUMBER >= 1070
+        return G4UIcommand::ConvertToLongInt(v.c_str());
+#else
         G4long vl;
         std::istringstream is(v);
         is >> vl;
         return vl;
-#else
-        return G4UIcommand::ConvertToLongInt(v.c_str());
 #endif
     }
 };

--- a/src/accel/SetupOptionsMessenger.cc
+++ b/src/accel/SetupOptionsMessenger.cc
@@ -10,6 +10,7 @@
 #include <type_traits>
 #include <G4UIcmdWithABool.hh>
 #include <G4UIcmdWithAnInteger.hh>
+#include <G4Version.hh>
 
 #include "corecel/Assert.hh"
 #include "corecel/sys/Device.hh"
@@ -54,7 +55,15 @@ struct UICommandTraits<T, std::enable_if_t<std::is_integral_v<T>>>
     static std::string to_string(T v) { return std::to_string(v); }
     static long from_string(G4String const& v)
     {
+        // Conversion to long int introduced in Geant4 10.7.0
+#if G4VERSION_NUMBER < 1070
+        G4long vl;
+        std::istringstream is(v);
+        is >> vl;
+        return vl;
+#else
         return G4UIcommand::ConvertToLongInt(v.c_str());
+#endif
     }
 };
 

--- a/src/celeritas/ext/g4vg/SolidConverter.cc
+++ b/src/celeritas/ext/g4vg/SolidConverter.cc
@@ -386,7 +386,10 @@ auto SolidConverter::orb(arg_type solid_base) -> result_type
 auto SolidConverter::para(arg_type solid_base) -> result_type
 {
     auto const& solid = dynamic_cast<G4Para const&>(solid_base);
-#if G4VERSION_NUMBER < 1100
+#if G4VERSION_NUMBER >= 1100
+    double const theta = solid.GetTheta();
+    double const phi = solid.GetPhi();
+#else
     // Theta/Phi are not directly accessible before 11.0 but are encoded in the
     // symmetry axis vector in the normalized x, y components. Geant4
     // internally constructs the unnormalized vector with z = 1, so can be
@@ -401,9 +404,6 @@ auto SolidConverter::para(arg_type solid_base) -> result_type
         = std::atan(std::sqrt(tan_theta_cos_phi * tan_theta_cos_phi
                               + tan_theta_sin_phi * tan_theta_sin_phi));
     double const phi = std::atan2(tan_theta_sin_phi, tan_theta_cos_phi);
-#else
-    double const theta = solid.GetTheta();
-    double const phi = solid.GetPhi();
 #endif
     return GeoManager::MakeInstance<UnplacedParallelepiped>(
         this->convert_scale_(solid.GetXHalfLength()),
@@ -581,7 +581,12 @@ auto SolidConverter::torus(arg_type solid_base) -> result_type
 auto SolidConverter::trap(arg_type solid_base) -> result_type
 {
     auto const& solid = dynamic_cast<G4Trap const&>(solid_base);
-#if G4VERSION_NUMBER < 1100
+#if G4VERSION_NUMBER >= 1100
+    double const theta = solid.GetTheta();
+    double const phi = solid.GetPhi();
+    double const alpha_1 = solid.GetAlpha1();
+    double const alpha_2 = solid.GetAlpha2();
+#else
     // Theta/Phi are not directly accessible before 11.0 but are encoded in the
     // symmetry axis vector x, y components. Here they are multiplied by
     // cos(theta), which is the z component, and so we can reconstruct them.
@@ -597,11 +602,6 @@ auto SolidConverter::trap(arg_type solid_base) -> result_type
                               + tan_theta_sin_phi * tan_theta_sin_phi));
     double const alpha_1 = std::atan(solid.GetTanAlpha1());
     double const alpha_2 = std::atan(solid.GetTanAlpha2());
-#else
-    double const theta = solid.GetTheta();
-    double const phi = solid.GetPhi();
-    double const alpha_1 = solid.GetAlpha1();
-    double const alpha_2 = solid.GetAlpha2();
 #endif
 
     return GeoManager::MakeInstance<UnplacedTrapezoid>(

--- a/src/celeritas/ext/g4vg/SolidConverter.cc
+++ b/src/celeritas/ext/g4vg/SolidConverter.cc
@@ -52,6 +52,7 @@
 #include <G4Tubs.hh>
 #include <G4UnionSolid.hh>
 #include <G4VSolid.hh>
+#include <G4Version.hh>
 #include <VecGeom/management/GeoManager.h>
 #include <VecGeom/volumes/UnplacedAssembly.h>
 #include <VecGeom/volumes/UnplacedBooleanVolume.h>
@@ -385,13 +386,32 @@ auto SolidConverter::orb(arg_type solid_base) -> result_type
 auto SolidConverter::para(arg_type solid_base) -> result_type
 {
     auto const& solid = dynamic_cast<G4Para const&>(solid_base);
+#if G4VERSION_NUMBER < 1100
+    // Theta/Phi are not directly accessible before 11.0 but are encoded in the
+    // symmetry axis vector in the normalized x, y components. Geant4
+    // internally constructs the unnormalized vector with z = 1, so can be
+    // reconstituted
+    G4ThreeVector const axis = solid.GetSymAxis();
+    double const original_mag = 1.0 / axis.z();
+    double const tan_theta_cos_phi = axis.x() * original_mag;
+    double const tan_theta_sin_phi = axis.y() * original_mag;
+
+    // Calculation taken from GetTheta/Phi implementations in Geant4 11
+    double const theta
+        = std::atan(std::sqrt(tan_theta_cos_phi * tan_theta_cos_phi
+                              + tan_theta_sin_phi * tan_theta_sin_phi));
+    double const phi = std::atan2(tan_theta_sin_phi, tan_theta_cos_phi);
+#else
+    double const theta = solid.GetTheta();
+    double const phi = solid.GetPhi();
+#endif
     return GeoManager::MakeInstance<UnplacedParallelepiped>(
         this->convert_scale_(solid.GetXHalfLength()),
         this->convert_scale_(solid.GetYHalfLength()),
         this->convert_scale_(solid.GetZHalfLength()),
         std::atan(solid.GetTanAlpha()),
-        solid.GetTheta(),
-        solid.GetPhi());
+        theta,
+        phi);
 }
 
 //---------------------------------------------------------------------------//
@@ -561,18 +581,41 @@ auto SolidConverter::torus(arg_type solid_base) -> result_type
 auto SolidConverter::trap(arg_type solid_base) -> result_type
 {
     auto const& solid = dynamic_cast<G4Trap const&>(solid_base);
+#if G4VERSION_NUMBER < 1100
+    // Theta/Phi are not directly accessible before 11.0 but are encoded in the
+    // symmetry axis vector x, y components. Here they are multiplied by
+    // cos(theta), which is the z component, and so we can reconstruct them.
+    G4ThreeVector const axis = solid.GetSymAxis();
+    double const tan_theta_cos_phi = axis.x() / axis.z();
+    double const tan_theta_sin_phi = axis.y() / axis.z();
+
+    // Calculation of theta, phi, alpha1, alpha2 using implementations as
+    // in Geant4 11
+    double const theta = std::atan2(tan_theta_sin_phi, tan_theta_cos_phi);
+    double const phi
+        = std::atan(std::sqrt(tan_theta_cos_phi * tan_theta_cos_phi
+                              + tan_theta_sin_phi * tan_theta_sin_phi));
+    double const alpha_1 = std::atan(solid.GetTanAlpha1());
+    double const alpha_2 = std::atan(solid.GetTanAlpha2());
+#else
+    double const theta = solid.GetTheta();
+    double const phi = solid.GetPhi();
+    double const alpha_1 = solid.GetAlpha1();
+    double const alpha_2 = solid.GetAlpha2();
+#endif
+
     return GeoManager::MakeInstance<UnplacedTrapezoid>(
         this->convert_scale_(solid.GetZHalfLength()),
-        solid.GetTheta(),
-        solid.GetPhi(),
+        theta,
+        phi,
         this->convert_scale_(solid.GetYHalfLength1()),
         this->convert_scale_(solid.GetXHalfLength1()),
         this->convert_scale_(solid.GetXHalfLength2()),
-        solid.GetAlpha1(),
+        alpha_1,
         this->convert_scale_(solid.GetYHalfLength2()),
         this->convert_scale_(solid.GetXHalfLength3()),
         this->convert_scale_(solid.GetXHalfLength4()),
-        solid.GetAlpha2());
+        alpha_2);
 }
 
 //---------------------------------------------------------------------------//

--- a/src/celeritas/ext/g4vg/SolidConverter.cc
+++ b/src/celeritas/ext/g4vg/SolidConverter.cc
@@ -90,6 +90,27 @@
 #include "Scaler.hh"
 #include "Transformer.hh"
 
+namespace
+{
+/*!
+ * Return theta, phi angles for a G4Para or G4Trap given their symmetry axis
+ */
+auto calculate_theta_phi(G4ThreeVector const& axis) -> std::pair<double, double>
+{
+    // The components of the symmetry axis for G4Para/Trap are alway encoded
+    // as a vector (A.tan(theta)cos(phi), A.tan(theta)sin(phi), A).
+    double const tan_theta_cos_phi = axis.x() / axis.z();
+    double const tan_theta_sin_phi = axis.y() / axis.z();
+
+    // Calculation taken from GetTheta/Phi implementations in Geant4 11
+    double const theta
+        = std::atan(std::sqrt(tan_theta_cos_phi * tan_theta_cos_phi
+                              + tan_theta_sin_phi * tan_theta_sin_phi));
+    double const phi = std::atan2(tan_theta_sin_phi, tan_theta_cos_phi);
+    return {theta, phi};
+}
+}  // namespace
+
 using namespace vecgeom;
 
 namespace celeritas
@@ -391,19 +412,8 @@ auto SolidConverter::para(arg_type solid_base) -> result_type
     double const phi = solid.GetPhi();
 #else
     // Theta/Phi are not directly accessible before 11.0 but are encoded in the
-    // symmetry axis vector in the normalized x, y components. Geant4
-    // internally constructs the unnormalized vector with z = 1, so can be
-    // reconstituted
-    G4ThreeVector const axis = solid.GetSymAxis();
-    double const original_mag = 1.0 / axis.z();
-    double const tan_theta_cos_phi = axis.x() * original_mag;
-    double const tan_theta_sin_phi = axis.y() * original_mag;
-
-    // Calculation taken from GetTheta/Phi implementations in Geant4 11
-    double const theta
-        = std::atan(std::sqrt(tan_theta_cos_phi * tan_theta_cos_phi
-                              + tan_theta_sin_phi * tan_theta_sin_phi));
-    double const phi = std::atan2(tan_theta_sin_phi, tan_theta_cos_phi);
+    // symmetry axis
+    auto const [theta, phi] = calculate_theta_phi(solid.GetSymAxis());
 #endif
     return GeoManager::MakeInstance<UnplacedParallelepiped>(
         this->convert_scale_(solid.GetXHalfLength()),
@@ -588,18 +598,8 @@ auto SolidConverter::trap(arg_type solid_base) -> result_type
     double const alpha_2 = solid.GetAlpha2();
 #else
     // Theta/Phi are not directly accessible before 11.0 but are encoded in the
-    // symmetry axis vector x, y components. Here they are multiplied by
-    // cos(theta), which is the z component, and so we can reconstruct them.
-    G4ThreeVector const axis = solid.GetSymAxis();
-    double const tan_theta_cos_phi = axis.x() / axis.z();
-    double const tan_theta_sin_phi = axis.y() / axis.z();
-
-    // Calculation of theta, phi, alpha1, alpha2 using implementations as
-    // in Geant4 11
-    double const theta = std::atan2(tan_theta_sin_phi, tan_theta_cos_phi);
-    double const phi
-        = std::atan(std::sqrt(tan_theta_cos_phi * tan_theta_cos_phi
-                              + tan_theta_sin_phi * tan_theta_sin_phi));
+    // symmetry axis
+    auto const [theta, phi] = calculate_theta_phi(solid.GetSymAxis());
     double const alpha_1 = std::atan(solid.GetTanAlpha1());
     double const alpha_2 = std::atan(solid.GetTanAlpha2());
 #endif

